### PR TITLE
Add expected type

### PR DESCRIPTION
**Pull request by ericmj**

### CODE REVIEW

**Comment by ericmj** [comment]
@josevalim I will hold off on passing the expected type down for blocks that have clauses or conditional patterns until we know how to handle unions. I think that's where we ended up in our talk, correct?

**Comment by josevalim** [comment]
@ericmj yup, just add a TODO for now.

**Review by josevalim — APPROVED** [review]

**Comment by ericmj** [comment]
I added TODOs for the places where we know how we will use the expected type. I left it out in the other places because I can't write a good TODO if we don't know how to handle it yet.

Places where we might use it in the future:

* `<-` in `with` and `for`
* `case`
* `with`
* `for`
* `try`
* `receive`

**Comment by josevalim** [comment]
Your list is good. Just __block__ missing I think. But basically anywhere that handles blocks.

**Comment by ericmj** [comment]
Block is already implemented https://github.com/elixir-lang/elixir/pull/10459/files#diff-4b9f31154b7080efa898901fe2f698d4224edf10b8023206e07628193aeda01dR207.

**Comment by josevalim** [comment]
@ericmj if you did it for blocks, then I would do it for everything:

1. Introduce a `of_block()` expression. If it receives a `{:__block__, .., ...}` then you do what you are doing in block.
2. call of_block pretty much everywhere, except on `for` which is a bit more complicated
3. <- in for  and with can be either TODOed or handled

File: lib/elixir/lib/module/types/expr.ex [hunk 1]
```diff
@@ -160,55 +167,68 @@ defmodule Module.Types.Expr do
...
-  def of_expr({:__block__, _meta, []}, _stack, context) do
+  def of_expr({:__block__, _meta, []}, _expected, _stack, context) do
     {:ok, {:atom, nil}, context}
   end
 
   # (expr; expr)
-  def of_expr({:__block__, _meta, exprs}, stack, context) do
-    case map_reduce_ok(exprs, context, &of_expr(&1, stack, &2)) do
+  def of_expr({:__block__, _meta, exprs}, expected, stack, context) do
+    expected_types = for(_ <- 1..(length(exprs) - 1), do: :dynamic) ++ [expected]
```
Comment (josevalim) [inline lib/elixir/lib/module/types/expr.ex:208]:
```suggestion
    expected_types = List.duplicate(:dynamic, length(exprs) - 1) ++ [expected]
```

?

**Pull request merged**